drbd: 9.19.1 -> 9.27.0 (utils), 9.2.7 kernel driver added

birkb opened the pull request:
## Description of changes

- DRBD [userland utilities](https://github.com/LINBIT/drbd-utils) updated from `9.19.1` to `9.27.0`
- DRBD [out-of-tree kernel driver](https://github.com/LINBIT/drbd) added with version `9.2.7`
- package splitted into [utils.nix](https://github.com/birkb/nixpkgs/blob/drbd-kernel-module/pkgs/os-specific/linux/drbd/utils.nix) and [driver.nix](https://github.com/birkb/nixpkgs/blob/drbd-kernel-module/pkgs/os-specific/linux/drbd/driver.nix)
- small [deprecation fix](https://github.com/birkb/nixpkgs/commit/fb297bf63137f20d236403b8f0c9e24abcdfb193) for the package test
- package test has been successful
- `nixpkgs-review` was successful for all but the broken [testing_bcachefs](https://github.com/NixOS/nixpkgs/pull/268394) kernel
- Release Notes for `23.11` and `24.05` [updated](https://github.com/birkb/nixpkgs/commit/361df5ca372ae6274385cfd917bdf3a26312d0c8)
- added my Github user to the [maintainer list](https://github.com/birkb/nixpkgs/commit/bd205a784b9848c4023e9b387565a8e1540c6919)

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
    - `nix-build ${NIXPKGS} --attr pkgs.drbd.passthru.tests`
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
    - `nix-build --attr nixosTests.drbd`
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] `nix-build pkgs/top-level/release.nix -A tarball.nixpkgs-basic-release-checks --arg supportedSystems '[ "aarch64-linux" "x86_64-linux" ]'` 
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

astro commented on pkgs/top-level/all-packages.nix:
```diff
@@ -27780,7 +27780,7 @@ with pkgs;
   dmraid = callPackage ../os-specific/linux/dmraid { lvm2 = lvm2_dmeventd; };
 
-  drbd = callPackage ../os-specific/linux/drbd { };
+  drbd = callPackage ../os-specific/linux/drbd/utils.nix { };
```
Thanks for keeping the pkg name!

astro commented on pkgs/os-specific/linux/drbd/driver.nix:
```diff
@@ -0,0 +1,59 @@
+    install -D drbd/drbd_transport_tcp.ko -t $out/lib/modules/${kernel.modDirVersion}/kernel/drivers/block/drbd
+    install -D drbd/drbd_transport_lb-tcp.ko -t $out/lib/modules/${kernel.modDirVersion}/kernel/drivers/block/drbd
+    install -D drbd/drbd_transport_rdma.ko -t $out/lib/modules/${kernel.modDirVersion}/kernel/drivers/block/drbd
+    '' + lib.optionalString (lib.versionOlder kernel.version "6.4") ''
+    install -D drbd/drbd-kernel-compat/handshake/handshake.ko -t $out/lib/modules/${kernel.modDirVersion}/kernel/drivers/block/drbd
+    '' + ''
```
Personally, I favor string interpolation for more readability:

```suggestion
    ${lib.optionalString (lib.versionOlder kernel.version "6.4") ''
      install -D drbd/drbd-kernel-compat/handshake/handshake.ko -t $out/lib/modules/${kernel.modDirVersion}/kernel/drivers/block/drbd
    ''}
```

Janik-Haag reviewed (commented):
the maintainer entry looks good. but please follow our commit conventions and try keeping a atomic commit history. For example the commit adding your self as maintainer should be called `maintainers: add birkb` and things like package updates should be called `$package: $old_versoin -> $new_version`

Janik-Haag commented:
please try sticking to `->` instead of fancy unicode arrow symbols.

birkb commented:
> the maintainer entry looks good. but please follow our commit conventions and try keeping a atomic commit history. For example the commit adding your self as maintainer should be called `maintainers: add birkb` and things like package updates should be called `$package: $old_versoin -> $new_version`

Hi Janik, i have [tried](https://github.com/NixOS/nixpkgs/pull/278449#issuecomment-1876881859) it... :grin:

birkb commented:
> please try sticking to `->` instead of fancy unicode arrow symbols.

I have copied that from the `nixpkgs-review` review output from my terminal. I have to check what happens there...

astro commented on pkgs/os-specific/linux/drbd/driver.nix:
```diff
@@ -0,0 +1,59 @@
+       DRBD is a software-based, shared-nothing, replicated storage solution
+       mirroring the content of block devices (hard disks, partitions, logical volumes, and so on) between hosts.
+    '';
+    broken = kernel.version == "6.5.13-bcachefs-unstable-2023-06-28";
```
This is oddly specific. Where does this kernel version come from, and why would it be the only one that is broken? Maybe we can make this check more general.

birkb replied:
> This is oddly specific. Where does this kernel version come from, and why would it be the only one that is broken? Maybe we can make this check more general.

The [testing_bcachefs](https://github.com/NixOS/nixpkgs/pull/268394) kernel seems to be broken quite often and should be removed anyhow. I wanted to configure `meta.broken` with the generic `testing_bcachefs` name, but i have found no way to do that. Any ideas?

birkb replied:
`broken = !kernel.meta.broken;` is doing the job in a very generic way...

```
1 package marked as broken and skipped:
linuxKernel.packages.linux_testing_bcachefs.drbd

1 package blacklisted:
nixos-install-tools

23 packages built:
drbd linuxKernel.packages.linux_4_19.drbd linuxKernel.packages.linux_4_19_hardened.drbd linuxKernel.packages.linux_5_10.drbd linuxKernel.packages.linux_5_10_hardened.drbd linuxKernel.packages.linux_5_15.drbd linuxKernel.packages.linux_5_15_hardened.drbd linuxKernel.packages.linux_5_4.drbd linuxKernel.packages.linux_5_4_hardened.drbd linuxKernel.packages.linux_6_1.drbd linuxKernel.packages.linux_hardened.drbd linuxKernel.packages.linux_6_5.drbd linuxKernel.packages.linux_6_5_hardened.drbd linuxKernel.packages.linux_6_6.drbd linuxKernel.packages.linux_6_6_hardened.drbd linuxKernel.packages.linux_latest_libre.drbd linuxKernel.packages.linux_libre.drbd linuxKernel.packages.linux_lqx.drbd linuxKernel.packages.linux_xanmod.drbd linuxKernel.packages.linux_xanmod_latest.drbd linuxKernel.packages.linux_zen.drbd ocf-resource-agents pacemaker
```

astro replied:
Is that line needed at all since https://github.com/NixOS/nixpkgs/commit/39805096cb9226f357364bdf9cc5492a8429b2ce?

I don't see the relationship between drbd and bcachefs. Even with `broken = !kernel.meta.broken;` I still do not think the drbd package should care at all if the user chooses to build broken kernels. It is not done by any other driver package either.

birkb replied:
Yes, since https://github.com/NixOS/nixpkgs/pull/268394 has been accepted the problem has gone away. The intention for `broken = !kernel.meta.broken;` is to not let the local `nixpkgs-review` run fail because drbd cannot be build because of a broken kernel. This should also help the central [hydra](https://github.com/NixOS/hydra) builds, or?

astro replied:
If the problem has gone away, can we now drop this line?

birkb replied:
Yes, done.

astro commented on nixos/doc/manual/release-notes/rl-2405.section.md:
```diff
@@ -159,3 +159,7 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
 - QtMultimedia has changed its default backend to `QT_MEDIA_BACKEND=ffmpeg` (previously `gstreamer` on Linux or `darwin` on MacOS).
   The previous native backends remain available but are now minimally maintained. Refer to [upstream documentation](https://doc.qt.io/qt-6/qtmultimedia-index.html#ffmpeg-as-the-default-backend) for further details about each platform.
+
+- The `drbd` package has been updated from `9.19.1` to `9.27.0`.
```
The NixOS release notes don't need every version bump. The line below is appropriate however.

astro reviewed (changes requested):
I am going to approve once the commits follow the nixpkgs conventions. Try `git rebase -i` or `lazygit` for that.

astro commented:
The version still has a typo in the commit message.

birkb commented:
> The version still has a typo in the commit message.

I have fixed the typo.

Janik-Haag commented on nixos/doc/manual/release-notes/rl-2311.section.md:
```diff
@@ -1320,6 +1317,7 @@ Make sure to also check the many updates in the [Nixpkgs library](#sec-release-2
   The configuration file could previously be read, when `userControlled.enable` (non-default),
   by users who are in both `wheel` and `userControlled.group` (defaults to `wheel`)
 
+- The `drbd` out-of-tree Linux kernel driver has been added in version `9.2.7`. With it the DRBD 9.x features can be used instead of the 8.x features provided by the `8.4.11` in-tree driver.
```
please drop this change since 23.11 was already released.

birkb replied:
done

Janik-Haag commented on pkgs/os-specific/linux/drbd/driver.nix:
```diff
@@ -0,0 +1,58 @@
+stdenv.mkDerivation rec {
+  name = "drbd-${version}-${kernel.version}";
+  version = "9.2.7";
+  sha256 = "1355ns10z0fjgqsdpf09qfy01j8lg2n7zy4kclmar3s798n3mh56";
+
+  src = fetchurl {
+    url = "https://pkg.linbit.com//downloads/drbd/9/drbd-${version}.tar.gz";
+    inherit sha256;
+  };
```
```suggestion

  src = fetchurl {
    url = "https://pkg.linbit.com//downloads/drbd/9/drbd-${version}.tar.gz";
    sha256 = "1355ns10z0fjgqsdpf09qfy01j8lg2n7zy4kclmar3s798n3mh56";
  };
```

Janik-Haag replied:
is there a reason to not just write the hash in the fetcher?

birkb replied:
I just wanted to keep the variables in one place, but it is only used once and nearby...

Janik-Haag commented on maintainers/maintainer-list.nix:
```diff
@@ -2469,6 +2469,12 @@
     githubId = 30630233;
     name = "Timo Triebensky";
   };
+  birkb = {
+    email = "birk@batchworks.de";
+    github = "birkb";
+    githubId = 10164833;
+    name = "Birk Bohne";
+  };
```
please split this into it's own commit called `maintainers: add birkb`

birkb replied:
done

Janik-Haag commented on pkgs/os-specific/linux/drbd/utils.nix:
```diff
@@ -22,23 +24,24 @@
 
 stdenv.mkDerivation rec {
   pname = "drbd";
-  version = "9.19.1";
+  version = "9.27.0";
 
   src = fetchurl {
     url = "https://pkg.linbit.com/downloads/drbd/utils/${pname}-utils-${version}.tar.gz";
-    sha256 = "1l99kcrb0j85wxxmrdihpx9bk1a4sdi7wlp5m1x5l24k8ck1m5cf";
+    sha256 = "1qwdrjrgas8z8vc6c85xcrqaczjwyqd61yig01n44wa5z0j3v4aq";
   };
```
this commit should be called `drbd: 9.19.1 -> 9.27.0`

birkb replied:
done

Commit:
maintainers: add birkb

Signed-off-by: birkb <birk@batchworks.de>

Commit:
drbd: 9.19.1 -> 9.27.0

Signed-off-by: birkb <birk@batchworks.de>

Commit:
drbd kernel driver added
- out-of-tree kernel driver added with version 9.2.7
- package splitted into utils.nix and driver.nix
- small deprecation fix for the package test
- Release Notes for 24.05 updated

Signed-off-by: birkb <birk@batchworks.de>

birkb committed:
Merge branch 'master' into drbd-kernel-module

birkb committed:
Merge branch 'master' into drbd-kernel-module

birkb commented:
Hi @Janik-Haag and @astro can you please check the PR again. Thanks!

Janik-Haag commented:
I'm currently very thin on time, sorry. I can try to get to it next weekend or so.

birkb committed:
Merge branch 'master' into drbd-kernel-module

birkb commented:
Hi @Janik-Haag and @astro another reminder... Thanks!

Lassulus reviewed (approved):
looking good
